core/src/main/scala/com/lightbend/paradox/ErrorContext.scala

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright © 2015 - 2019 Lightbend, Inc. <http://www.lightbend.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.lightbend.paradox
+
+import java.io.File
+
+import com.lightbend.paradox.markdown.Page
+import org.pegdown.ast.Node
+
+import scala.collection.mutable.ListBuffer
+
+trait ErrorContext {
+  def apply(msg: String, index: Int): Unit
+
+  def apply(msg: String, node: Node): Unit = apply(msg, node.getStartIndex)
+
+  def apply(msg: String): Unit
+
+  def apply(msg: String, file: File, index: Int): Unit
+
+  def apply(msg: String, file: File, node: Node): Unit = apply(msg, file, node.getStartIndex)
+
+  def apply(msg: String, file: File): Unit
+
+  def apply(msg: String, page: Page, index: Int): Unit = apply(msg, page.file, index)
+
+  def apply(msg: String, page: Page, node: Node): Unit = apply(msg, page.file, node.getStartIndex)
+
+  def apply(msg: String, page: Page): Unit = apply(msg, page.file)
+}
+
+class ErrorCollector extends ErrorContext {
+  private val errors = new ListBuffer[ParadoxError]()
+
+  private def addError(msg: String, page: Option[File], index: Option[Int]) = {
+    errors.append(ParadoxError(msg, page, index))
+  }
+
+  override def apply(msg: String, index: Int): Unit = throw new IllegalArgumentException("Cannot report an indexed error without a page context")
+
+  override def apply(msg: String): Unit = addError(msg, None, None)
+
+  override def apply(msg: String, file: File, index: Int): Unit = addError(msg, Some(file), Some(index))
+
+  override def apply(msg: String, file: File): Unit = addError(msg, Some(file), None)
+
+  def hasErrors: Boolean = errors.nonEmpty
+
+  def errorCount: Int = errors.toList.distinct.size
+
+  def logErrors(log: ParadoxLogger) = {
+    val totalErrors = errors.toList.distinct
+    // First log general errors
+    totalErrors.foreach {
+      case ParadoxError(msg, None, _) => log.error(msg)
+      case _                          =>
+    }
+    val pageErrors = totalErrors.collect {
+      case ParadoxError(msg, Some(page), idx) => (msg, page, idx)
+    }
+    pageErrors.groupBy(_._2).toSeq.sortBy(_._1.getAbsolutePath).foreach {
```

Comment:
Minor readability comment that can be ignored: but I needed to go back and forth to confirm the tuples (such as `_2` and `_3` later). Using case-based anonymous functions to name could be clearer when reading through, or even something to retain the `ParadoxError`s like:

```
totalErrors
  .filter(_.page.isDefined)
  .groupBy(_.page.get)
  ...
```

Reply:
Updated as suggested.